src/DynamicRupture/Output/Base.cpp

```diff
@@ -214,77 +216,329 @@ void Base::updateElementwiseOutput() {
-    if (peakSlipsRate.isActive) {
-      peakSlipsRate(level, i) = peakSR[ltsId][nearestGaussPoint];
+    if (type == OutputType::AtPickpoint) {
+      outputData.cachedTime[outputData.currentCacheLevel] = time;
+      outputData.currentCacheLevel += 1;
     }
+  }
+}
+
+void Base::computeLocalStresses() {
+  real pFactorPlus = local.waveSpeedsPlus->pWaveVelocity * local.waveSpeedsPlus->density;
+  real pFactorMinus = local.waveSpeedsMinus->pWaveVelocity * local.waveSpeedsMinus->density;
+
+  real sFactorPlus = local.waveSpeedsPlus->sWaveVelocity * local.waveSpeedsPlus->density;
+  real sFactorMinus = local.waveSpeedsMinus->sWaveVelocity * local.waveSpeedsMinus->density;
+
+  real norDivisor = 1.0 / (pFactorMinus + pFactorPlus);
+  real shearDivisor = 1.0 / (sFactorMinus + sFactorPlus);
+  real uVelDivisor = 1.0 / (pFactorPlus);
+
+  auto diff = [this](int i) {
+    return this->local.faceAlignedValuesMinus[i] - this->local.faceAlignedValuesPlus[i];
+  };
+
+  local.xyStress = local.faceAlignedValuesPlus[3] +
+                   ((diff(3) + sFactorMinus * diff(7)) * sFactorPlus) * shearDivisor;
+
+  local.xzStress = local.faceAlignedValuesPlus[5] +
+                   ((diff(5) + sFactorMinus * diff(8)) * sFactorPlus) * shearDivisor;
+
+  local.p = local.faceAlignedValuesPlus[0] +
+            ((diff(0) + pFactorMinus * diff(6)) * pFactorPlus) * norDivisor;
+
+  local.u =
+      local.faceAlignedValuesPlus[6] + (local.p - local.faceAlignedValuesPlus[0]) * uVelDivisor;
+
+  real missingSigmaValues = (local.p - local.faceAlignedValuesPlus[0]);
+  missingSigmaValues *= (1.0 - 2.0 * std::pow(local.waveSpeedsPlus->sWaveVelocity /
+                                                  local.waveSpeedsPlus->pWaveVelocity,
+                                              2));
+
+  local.yyStress = local.faceAlignedValuesPlus[1] + missingSigmaValues;
+  local.zzStress = local.faceAlignedValuesPlus[2] + missingSigmaValues;
+  local.yzStress = local.faceAlignedValuesPlus[4];
+
+  local.tracEla =
+      std::sqrt(std::pow(local.sXY + local.xyStress, 2) + std::pow(local.sXZ + local.xzStress, 2));
+}
 
-    auto& slipVectors = std::get<VariableID::Slip>(outputData.vars);
-    if (slipVectors.isActive) {
-      VrtxCoords crossProduct = {0.0, 0.0, 0.0};
-      MeshTools::cross(strike, tangent1, crossProduct);
+void Base::computeLocalTraction(real strength) {
+  if (local.tracEla > std::abs(strength)) {
+    local.xyTraction = ((local.sXY + local.xyStress) / local.tracEla) * strength;
+    local.xzTraction = ((local.sXZ + local.xzStress) / local.tracEla) * strength;
 
-      double cos1 = MeshTools::dot(strike, tangent1);
-      double scalarProd = MeshTools::dot(crossProduct, normal);
+    // update stress change
+    local.xyTraction -= local.sXY;
+    local.xzTraction -= local.sXZ;
+  } else {
+    local.xyTraction = local.xyStress;
+    local.xzTraction = local.xzStress;
+  }
+}
+
+void Base::computeSlipAndRate(std::array<real, 6>& rotatedTraction,
```

Comment:
actually, this should be named computeSlipRate?

Reply:
ranamed `computeSlipAndRate` to `computeSlipRate`